### PR TITLE
#69 filter items with pagination

### PR DESCRIPTION
**Pull request by laganows**

### CODE REVIEW

File: src/main/java/com/beerjournal/breweriana/collection/persistence/UserCollectionRepository.java [hunk 1]
```diff
@@ -29,18 +36,35 @@
...
+                list.stream().filter(v -> filterVariableName.equals("name") ?
+                        v.getName().startsWith(filterVariableValue) : v.getType().startsWith(filterVariableValue));
+
+        return new PageImpl<>(
+                listStream
+                        .skip(page * count)
+                        .limit(count)
+                        .collect(Collectors.toList()),
+                new PageRequest(page, count),
+                list.size());
```
Comment (laganows) [inline src/main/java/com/beerjournal/breweriana/collection/persistence/UserCollectionRepository.java]: I believe that there is no need to use crudRepostory (need to query all items to find them for given userId) for this operation and stream on given list is enough. It shouldn't be a problem with filtering. Am I right?

File: src/main/java/com/beerjournal/breweriana/collection/persistence/UserCollectionRepository.java [hunk 2]
```diff
@@ -85,15 +109,21 @@ int updateItem(Item item) {
...
-                new Query(Criteria.where("name").not().in(userItemsNames)),
-                Item.class);
+    private Page<Item> findAllNotIn(Set<String> userItemsNames, PageRequest pageRequest, String filterVariableName, String filterVariableValue) {
+        Criteria criteria = Criteria.where("name").not().in(userItemsNames);
+        if (!Strings.isNullOrEmpty(filterVariableValue))
+            criteria = criteria.andOperator(Criteria.where(filterVariableName).regex(Pattern.compile("^" + filterVariableValue)));
+
+        Query query = new Query(criteria);
+        long total = mongoOperations.count(query, Item.class);
+        List<Item> items = mongoOperations.find(query.with(pageRequest), Item.class);
```
Comment (laganows) [inline src/main/java/com/beerjournal/breweriana/collection/persistence/UserCollectionRepository.java:125]: Problem with pagination with query (http://stackoverflow.com/questions/29030542/pagination-with-mongotemplate).

So I workaround it with http://stackoverflow.com/questions/20366666/spring-data-mongo-pagination (phuong answear.)

File: src/main/java/com/beerjournal/breweriana/collection/persistence/UserCollectionRepository.java [hunk 3]
```diff
@@ -29,18 +36,35 @@
         return crudRepository.findOneByOwnerId(ownerId);
     }
 
-    public Set<ItemRef> findAllNotInUserCollection(ObjectId ownerId) {
+    public Page<ItemRef> findAllInUserCollection(ObjectId ownerId, int page, int count, String filterVariableName, String filterVariableValue) {
+        UserCollection userCollection = crudRepository.findOneByOwnerId(ownerId)
+                .orElseThrow(() -> new BeerJournalException(USER_COLLECTION_NOT_FOUND));
+        List<ItemRef> list = new ArrayList<>(userCollection.getItemRefs());
```
Comment (pkruczek) [inline src/main/java/com/beerjournal/breweriana/collection/persistence/UserCollectionRepository.java]: "list" is not proper name

Comment (laganows) [reply src/main/java/com/beerjournal/breweriana/collection/persistence/UserCollectionRepository.java]: Changed to userItems

**Review by pkruczek — APPROVED** [review]

**Pull request merged**